src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsLangProjectSystemService.cs

```diff
@@ -0,0 +1,287 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio.ComponentModelHost;
+using NuGet.Commands;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+using NuGet.Versioning;
+using NuGet.VisualStudio;
+using VSLangProj150;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Contains the information specific to a Visual Basic or C# project.
+    /// </summary>
+    internal class VsLangProjectSystemService
+        : IProjectSystemCapabilities
+        , IProjectSystemReferencesReader
+        , IProjectSystemReferencesService
+    {
+        private static readonly Array _referenceMetadata;
+
+        private readonly EnvDTE.Project _project;
+        private readonly IVsProjectThreadingService _threadingService;
+
+        private readonly Lazy<VSProject4> _asVSProject4;
+
+        private VSProject4 AsVSProject4 => _asVSProject4.Value;
+
+        public bool SupportsPackageReferences => true;
+
+        static VsLangProjectSystemService()
+        {
+            _referenceMetadata = Array.CreateInstance(typeof(string), 3);
+            _referenceMetadata.SetValue(ProjectItemProperties.IncludeAssets, 0);
+            _referenceMetadata.SetValue(ProjectItemProperties.ExcludeAssets, 1);
+            _referenceMetadata.SetValue(ProjectItemProperties.PrivateAssets, 2);
+        }
+
+        public VsLangProjectSystemService(
+            IVsProjectAdapter projectAdapter,
+            INuGetProjectServices projectServices)
+        {
+            Assumes.Present(projectAdapter);
+
+            _project = projectAdapter.Project;
+
+            _threadingService = projectServices.GetGlobalService<IVsProjectThreadingService>();
+            Assumes.Present(_threadingService);
+
+            _asVSProject4 = new Lazy<VSProject4>(() => _project.Object as VSProject4);
```

Comment:
We use Lazy freely wherever it's possible. The reason here is to defer accessing shell object until it's really needed and when switching to the main thread will be possible (not in the ctor).